backend/authschemes/localauth/local_auth.go

```diff
@@ -56,39 +73,68 @@ func (LocalAuthScheme) FriendlyName() string {
 // the underlying system/database
 func (p LocalAuthScheme) BindRoutes(r *mux.Router, bridge authschemes.AShirtAuthBridge) {
 	remux.Route(r, "POST", "/register", remux.JSONHandler(func(r *http.Request) (interface{}, error) {
+		if !p.RegistrationEnabled {
+			return nil, fmt.Errorf("registration is closed to users")
+		}
+
 		dr := remux.DissectJSONRequest(r)
-		firstName := dr.FromBody("firstName").Required().AsString()
-		lastName := dr.FromBody("lastName").Required().AsString()
-		email := dr.FromBody("email").Required().AsString()
-		password := dr.FromBody("password").Required().AsString()
+		info := RegistrationInfo{
+			Email:     dr.FromBody("email").Required().AsString(),
+			FirstName: dr.FromBody("firstName").Required().AsString(),
+			LastName:  dr.FromBody("lastName").Required().AsString(),
+			Password:  dr.FromBody("password").Required().AsString(),
+		}
+
 		if dr.Error != nil {
 			return nil, dr.Error
 		}
 
-		if err := checkPasswordComplexity(password); err != nil {
+		if err := checkPasswordComplexity(info.Password); err != nil {
 			return nil, err
 		}
 
-		encryptedPassword, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
-		if err != nil {
-			return nil, backend.WrapError("Unable to generate encrypted password", err)
+		return nil, registerNewUser(r.Context(), bridge, info)
+	}))
+
+	remux.Route(r, "POST", "/admin/register", remux.JSONHandler(func(r *http.Request) (interface{}, error) {
+		if !middleware.IsAdmin(r.Context()) {
+			return nil, backend.UnauthorizedWriteErr(fmt.Errorf("Requesting user is not an admin"))
 		}
 
-		userResult, err := bridge.CreateNewUser(authschemes.UserProfile{
-			FirstName: firstName,
-			LastName:  lastName,
-			Slug:      strings.ToLower(firstName + "." + lastName),
-			Email:     email,
-		})
-		if err != nil {
+		authKey := make([]byte, 40)
+		if _, err := rand.Read(authKey); err != nil {
+			return nil, backend.WrapError("Unable to generate random new user password key", err)
+		}
+		// convert authKey into readable range (33-126)
+		readKey := make([]byte, len(authKey))
+		for i, b := range authKey {
+			readKey[i] = (b % (126 - 33)) + 33
```

Comment:
Why are we using mod to get the value into the ascii printable keyspace rather than just encoding to something printable (hex or base64)? Isn't this reducing the keyspace?

Reply:
My thinking at the time was that it was important to cast these values into a printable range, and having a larger set range of characters was better than going with base64 (after all: this range is 93 possible characters, rather than the 64 we'd get otherwise)

But thinking about this now, I think that's flawed. I think the entropy should be the same for the 40 characters represented as base64 as they would be in any other representation. Also, I'm not confident that the above mapping maps equally over all of the 93 possible characters, so the base64 approach might be better after all. I can revise into that approach if you agree.